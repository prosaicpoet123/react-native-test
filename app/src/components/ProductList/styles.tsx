import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexWrap: 'wrap',
    justifyContent: 'flex-start',
    marginLeft: 8,
    marginRight: 8,
  },
  item: {
    width: '50%',
    padding: 3,
  },
  loading: {
    justifyContent: 'center',
    alignItems: 'center',
    height: '100%',
  },
})

export default styles
