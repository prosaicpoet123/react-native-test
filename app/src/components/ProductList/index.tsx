import React, { FC, useEffect, useState } from 'react'
import { IProduct } from '../../@Types/product'
import { RootStackParamList } from '../../@Types/navigation'
import { ActivityIndicator, FlatList, Platform, Text, View } from 'react-native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import ProductCard from '../ProductCard'
import styles from './styles'

type IProps = NativeStackScreenProps<RootStackParamList, 'Product List'>

const ProductList: FC<IProps> = ({ navigation }) => {
  const [isLoading, setLoading] = useState(true)
  const [isError, setError] = useState(false)
  const [data, setData] = useState([] as IProduct[])
  const [offset, setOffset] = useState(1)

  useEffect(() => {
    const getProducts = async () => {
      try {
        // hit an issue with data fetching on Android emulator, this issue: https://stackoverflow.com/questions/33704130/react-native-android-fetch-failing-on-connection-to-local-api
        // localhost works fine on IOS, but for Android development I had to switch localhost to 10.0.2.2, for the purposes of local development this was okay
        // almost certain there is a better fix
        const response = await fetch(
          `http://${
            Platform.OS === 'ios' ? 'localhost' : '10.0.2.2'
          }:4000/products/all?page=${offset}`,
        )
        const json = await response.json()
        setData((d) => [...d, ...json.products])
      } catch (error) {
        console.error(error)
        setError(true)
      } finally {
        setLoading(false)
      }
    }

    getProducts()
  }, [offset])

  if (isError) return <Text>Oops! Something Went Wrong...</Text>

  return (
    <View>
      {isLoading ? (
        <View style={styles.loading}>
          <ActivityIndicator size="large" color="#0000ff" />
        </View>
      ) : (
        <View style={styles.container}>
          <FlatList
            numColumns={2}
            horizontal={false}
            keyExtractor={(item, index) => `${item.id}-${index}`}
            data={data}
            renderItem={({ item }) => (
              <View style={styles.item}>
                <ProductCard
                  product={item}
                  onPress={(product: IProduct) =>
                    navigation.navigate('Product Details', { product })
                  }
                />
              </View>
            )}
            onEndReachedThreshold={20}
            onEndReached={() => setOffset(offset + 1)}
          />
        </View>
      )}
    </View>
  )
}

export default ProductList
