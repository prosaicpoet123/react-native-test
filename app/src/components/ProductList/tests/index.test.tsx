import React from 'react'
import renderer, { act } from 'react-test-renderer'
import ProductList from '..'

describe('<ProductList />', () => {
  it('renders correctly', async () => {
    const fakeProducts = {
      products: [
        {
          id: '1',
          name: 'Fake Product 1',
        },
        {
          id: '2',
          name: 'Fake Product 2',
        },
        {
          id: '3',
          name: 'Fake Product 3',
        },
      ],
    }

    global.fetch = jest.fn().mockImplementationOnce(() => {
      return new Promise((resolve) => {
        resolve({
          ok: true,
          json: () => fakeProducts,
        })
      })
    })

    let tree

    await act(async () => {
      tree = renderer.create(<ProductList />).toJSON()
    })

    expect(fetch).toHaveBeenCalledTimes(1)
    expect(tree).toMatchSnapshot()
  })
})
