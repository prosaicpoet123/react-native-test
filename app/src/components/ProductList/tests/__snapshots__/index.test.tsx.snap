// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductList /> renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "display": "flex",
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginLeft": 8,
        "marginRight": 8,
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Object {
            "id": "1",
            "images": Array [],
            "name": "Fake Product 1",
            "price": 20,
            "promotions": Array [],
            "rating": 1,
            "review_count": 30,
            "stock_status": "INSTOCK",
          },
          Object {
            "id": "2",
            "images": Array [],
            "name": "Fake Product 2",
            "price": 20,
            "promotions": Array [],
            "rating": 1,
            "review_count": 30,
            "stock_status": "OUTOFSTOCK",
          },
          Object {
            "id": "3",
            "images": Array [],
            "name": "Fake Product 3",
            "price": 20,
            "promotions": Array [],
            "rating": 1,
            "review_count": 30,
            "stock_status": "LOWSTOCK",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={25}
      keyExtractor={[Function]}
      maxToRenderPerBatch={25}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "padding": 3,
                  "width": "50%",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#FFFFFF",
                        "borderColor": "#CECECE",
                        "borderWidth": 1,
                        "minHeight": 300,
                        "padding": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  <View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <Image
                        source={
                          Object {
                            "uri": "https://images.hollandandbarrettimages.co.uk/productimages/HB/180/1_A.jpg",
                          }
                        }
                        style={
                          Array [
                            Object {
                              "height": 180,
                              "width": "100%",
                            },
                            Object {
                              "height": 180,
                            },
                          ]
                        }
                      />
                    </View>
                  </View>
                  <View
                    style={
                      Object {
                        "padding": 4,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "marginBottom": 8,
                          },
                          Object {
                            "fontWeight": "bold",
                          },
                          Object {
                            "fontSize": 14,
                          },
                        ]
                      }
                    >
                      Fake Product 1
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "padding": 3,
                  "width": "50%",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#FFFFFF",
                        "borderColor": "#CECECE",
                        "borderWidth": 1,
                        "minHeight": 300,
                        "padding": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  <View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <Image
                        source={
                          Object {
                            "uri": "https://images.hollandandbarrettimages.co.uk/productimages/HB/180/2_A.jpg",
                          }
                        }
                        style={
                          Array [
                            Object {
                              "height": 180,
                              "width": "100%",
                            },
                            Object {
                              "height": 180,
                            },
                          ]
                        }
                      />
                    </View>
                  </View>
                  <View
                    style={
                      Object {
                        "padding": 4,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "marginBottom": 8,
                          },
                          Object {
                            "fontWeight": "bold",
                          },
                          Object {
                            "fontSize": 14,
                          },
                        ]
                      }
                    >
                      Fake Product 2
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "padding": 3,
                  "width": "50%",
                }
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#FFFFFF",
                        "borderColor": "#CECECE",
                        "borderWidth": 1,
                        "minHeight": 300,
                        "padding": 4,
                      },
                      undefined,
                    ]
                  }
                >
                  <View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <Image
                        source={
                          Object {
                            "uri": "https://images.hollandandbarrettimages.co.uk/productimages/HB/180/3_A.jpg",
                          }
                        }
                        style={
                          Array [
                            Object {
                              "height": 180,
                              "width": "100%",
                            },
                            Object {
                              "height": 180,
                            },
                          ]
                        }
                      />
                    </View>
                  </View>
                  <View
                    style={
                      Object {
                        "padding": 4,
                      }
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "marginBottom": 8,
                          },
                          Object {
                            "fontWeight": "bold",
                          },
                          Object {
                            "fontSize": 14,
                          },
                        ]
                      }
                    >
                      Fake Product 3
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
